063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,221 @@
+< User-Task-ID: 5344ca89-351f-4565-8d0f-9aade00e053d
+< Content-Type: application/json;charset=utf-8
+< Cruise-Control-Version: 2.5.77-SNAPSHOT
+< Cruise-Control-Commit_Id: 535ea07e8847990cc02dde1f1be99b387dbeaf5b
+< Content-Length: 17051
+< Server: Jetty(9.4.52.v20230823)
+< 
+{ [17321 bytes data]
+{
+  "summary": {
+    "numIntraBrokerReplicaMovements": 0,
+    "numReplicaMovements": 125,
+    "onDemandBalancednessScoreAfter": 100,
+    "intraBrokerDataToMoveMB": 0,
+    "monitoredPartitionsPercentage": 100,
+    "provisionRecommendation": "",
+    "excludedBrokersForReplicaMove": [],
+    "excludedBrokersForLeadership": [],
+    "provisionStatus": "RIGHT_SIZED",
+    "onDemandBalancednessScoreBefore": 100,
+    "recentWindows": 5,
+    "dataToMoveMB": 0,
+    "excludedTopics": [],
+    "numLeaderMovements": 0
+  }
+  ...
+```
+
+The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.goals` property.
+
+The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.replication.throttle` property.
+
+The `skip_rack_awareness_check` boolean parameter is undocumented and disabled by default.
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
+
+#### User tasks endpoint
+
+The `user_tasks` endpoint is used to periodically check the tasks state by `User-Task-ID`, which is returned in the `topic_configuration` response.
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Topic Operator
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic also detects `.spec.replicas` changes by comparing with the actual topic replication factor in Kafka.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- **Pending**: replicas is updated, but the replication factor change request is not sent to Cruise Control.
+- **Ongoing**: replicas is updated and Cruise Control has accepted the request, but the execution is not completed.
+- **Ready**: replicas is updated and Cruise Control execution is completed.
```

Comment:
What do these correspnd to in terms of the Cruise Cotnrol states? Is Pending a task that is Active but not in execution yet? Or is Pending the state when the CC queue is full and the tasks is not queued yet? I guess the later?

Reply:
> Is Pending a task that is Active but not in execution yet?

No, that's the Ongoing state. Or better, an ongoing change is when the request has been accepted by CC (it's in the queue). In terms of CC states, it can be Active or InExecution. Proposal updated.

> is Pending the state when the CC queue is full and the tasks is not queued yet?

In that case, the Topic Operator gets an `OngoingExecutionException`, updates the KafkaTopic status with an appropriate message, waits for some time (backoff period), and then removes the KT from pending changes. The last operation allows a new notification from the periodic reconciliation to come in, which is basically a retry.

I added this example to the proposal.